src/zcash/NoteEncryption.cpp

```diff
@@ -0,0 +1,169 @@
+    BOOST_STATIC_ASSERT(32 == crypto_scalarmult_SCALARBYTES);
+    BOOST_STATIC_ASSERT(NOTEENCRYPTION_AUTH_BYTES == crypto_aead_chacha20poly1305_ABYTES);
+
+    // Create the ephemeral keypair
+    esk = random_uint256();
+    epk = generate_pubkey(esk);
+}
+
+template<size_t MLEN>
+NoteDecryption<MLEN>::NoteDecryption(uint256 sk_enc) : sk_enc(sk_enc) {
+    this->pk_enc = NoteEncryption<MLEN>::generate_pubkey(sk_enc);
+}
+
+template<size_t MLEN>
+typename NoteEncryption<MLEN>::Ciphertext NoteEncryption<MLEN>::encrypt
+                                          (const uint256 &pk_enc,
+                                           const NoteEncryption<MLEN>::Plaintext &message
+                                          )
+{
+    uint256 dhsecret;
+
+    if (crypto_scalarmult(dhsecret.begin(), esk.begin(), pk_enc.begin()) != 0) {
+        throw std::logic_error("Could not create DH secret");
+    }
+
+    // Construct the symmetric key
+    unsigned char K[NOTEENCRYPTION_CIPHER_KEYSIZE];
+    KDF(K, dhsecret, epk, pk_enc, hSig, nonce);
+
+    // Increment the number of encryptions we've performed
+    nonce++;
+
+    // The nonce is zero because we never reuse keys
+    unsigned char cipher_nonce[crypto_aead_chacha20poly1305_IETF_NPUBBYTES] = {};
+
+    NoteEncryption<MLEN>::Ciphertext ciphertext;
+
+    crypto_aead_chacha20poly1305_ietf_encrypt(ciphertext.begin(), NULL,
+                                         message.begin(), MLEN,
+                                         NULL, 0, // no "additional data"
+                                         NULL, cipher_nonce, K);
+
+    return ciphertext;
+}
+
+template<size_t MLEN>
+typename NoteDecryption<MLEN>::Plaintext NoteDecryption<MLEN>::decrypt
+                                         (const NoteDecryption<MLEN>::Ciphertext &ciphertext,
+                                          const uint256 &epk,
+                                          const uint256 &hSig,
+                                          unsigned char nonce
+                                         ) const
+{
+    uint256 dhsecret;
+
+    if (crypto_scalarmult(dhsecret.begin(), sk_enc.begin(), epk.begin()) != 0) {
+        throw std::logic_error("Could not create DH secret");
+    }
+
+    unsigned char K[NOTEENCRYPTION_CIPHER_KEYSIZE];
+    KDF(K, dhsecret, epk, pk_enc, hSig, nonce);
+
+    // The nonce is zero because we never reuse keys
+    unsigned char cipher_nonce[crypto_aead_chacha20poly1305_IETF_NPUBBYTES] = {};
+
+    NoteDecryption<MLEN>::Plaintext plaintext;
+
+    if (crypto_aead_chacha20poly1305_ietf_decrypt(plaintext.begin(), NULL,
+                                             NULL,
+                                             ciphertext.begin(), NoteDecryption<MLEN>::CLEN,
+                                             NULL,
+                                             0,
+                                             cipher_nonce, K) != 0) {
+        throw std::runtime_error("Could not decrypt message");
+    }
```

Comment:
Boost docs say "Elements may have an undetermined initial value" so this could potentially display uninitialized memory in the memo field or something like that.